### ISSUE TITLE
Encrypted Media Extensions

### ISSUE THREAD

**Issue by cadubentzen** [annulen/webkit#658]
In the WPE port, there already is support for Encrypted Media, and also some improvements for MSE using GStreamer.

Is the merging of this ongoing? In the wiki page it says help is needed for Encrypted Media. Besides merging WebCore, what other modifications are needed in the Qt port?

Maybe adding a new setting to QWebSettings?

In case no one has done it already, I could get start on it in my fork.

**Comment by annulen** [comment]
In fact there is some initial support EME in current branch, but I don't know if it's workable. So I think prerequisite is reaching https://github.com/annulen/webkit/milestone/5 when we'll have fresh WebKit from trunk with latest and greatest MSE/EME code, including GStreamer parts.

What can you do right now to help: research information about key systems which can be used by WPE and GTK ports. I know there is ClearKey support in trunk, but it is not a real DRM. What key systems do we need to support to access content on sites that use EME? What sites do GTK and WPE are actually targeting? Do we need to load binary modules for that, e.g. from Firefox, to get other key systems supported? Is there any out-of-trunk code for other key systems, e.g. in Metrological fork?

@eocanha Could you please sched some light on this topic?

If we need to load some kind of external modules we'll need API for that. Adding on/off switch for EME itself is trivial.

P.S. I don't have Netflix account, and I don't know if there is any public place where to test EME without paying money first.

**Comment by eocanha** [comment]
CCing @calvaris, as he might be interested in this thread.

The current upstream EME support doesn't work without some sort of proprietary CDM (which is not upstream for legal reasons), but it's the base of Metrological code, where a Playready CDM is available privately (legal reasons again, we're sorry about that).

Our team is currently working on EMEv3 and OpenCDM. We have hope that this OpenCDM work has more chances to be upstreamable and usable by more people, but I can't give any guarantee at this moment.

A proper way to test EME in a system with all the relevant pieces is to purchase some cheap movie in Google Play Video (price starts from 3.99€) and watch it via https://youtube.com/tv.

**Comment by calvaris** [comment]
I agree with @eocanha and just a small comment to complement what he said. We are working on EMEv3 with OpenCDM and we will probably upstream that code at some point but it will be never enabled by default because for OpenCDM to work we need the proprietary CDM components for the OpenCDM backends to work) and those need licenses. OpenCDM API has a compatible license so that we can write WebKit code against it but it is not usable at runtime because of the legal reasons. For this, we won't enable OpenCDM by default.

Considering that OpenCDM won't be enabled by default, then it does not make sense that we enable EME support by default because we would have ClearKey only and that's the reference toy implementation.

**Comment by annulen** [comment]
@eocanha @calvaris Thank you very much for the info!

As for OpenCDM support in trunk, can it be supported as a plugin? E.g. with NPAPI or custom plugin interface? In this case desktop users will be able to add EME support to normal binary packages. (Bonus points if same plugin could be loaded into WebKitGTK and QtWebKit :)

**Comment by annulen** [comment]
>for OpenCDM to work we need the proprietary CDM components for the OpenCDM backends to work

Is OpenCDM built against those components or they can be dynamically loaded?

**Comment by calvaris** [comment]
Answering both things, AFAIK, OpenCDM backends go thru RPC so we wouldn't need web plugins support to get them working if you have them installed. Therefore I guess there would be no technological problem to have them in WebKitGTK+, WPE and QtWebKit.

The big issue here is the licensing issues. We can release the use of OpenCDM code in WebKit, the OpenCDM code itself, including the backends but not the code those backends use because they are privative. Vendors will have to get licenses for those.

**Comment by annulen** [comment]
@calvaris Got it. So did you mean that you will disable it at run time (because it won't work on unprepared system), but enable at compile time?

**Comment by calvaris** [comment]
By default I'd disable at compile time and let people enable it at compile and runtime if they have the ability to install those things. IMHO, it does not make sense to have it enable in compile time if there is not chance to get it activated in runtime.

**Comment by annulen** [comment]
Well, at least widevine CDM is not hard to install for anyone, see e.g. https://aur.archlinux.org/packages/chromium-widevine/

**Comment by calvaris** [comment]
Might be. The problem here is that we depend on a customized version of OpenCDM that we'll try to upstream ASAP, but it will take time.
